x/bank/simulation/msgs.go

```diff
@@ -15,56 +15,84 @@ import (
-			initFromCoins[denomIndex].Denom,
-			toAddr.String(),
-		)
+		return action, nil, nil
+	}
+}
 
-		coins := sdk.Coins{{initFromCoins[denomIndex].Denom, amt}}
-		var msg = bank.MsgSend{
-			Inputs:  []bank.Input{bank.NewInput(fromAcc.Address, coins)},
-			Outputs: []bank.Output{bank.NewOutput(toAddr, coins)},
+// SingleInputSendMsg tests and runs a single msg send, with one input and one output, where both
+// accounts already exist.
+func SingleInputSendMsg(mapper auth.AccountMapper, bk bank.Keeper) simulation.Operation {
+	handler := bank.NewHandler(bk)
+	return func(r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simulation.Account, event func(string)) (action string, fOps []simulation.FutureOperation, err error) {
+		fromAcc, action, msg, abort := createSingleInputSendMsg(r, ctx, accs, mapper)
+		if abort {
+			return action, nil, nil
 		}
-		goErr = sendAndVerifyMsgSend(app, mapper, msg, ctx, []crypto.PrivKey{fromAcc.PrivKey})
-		if goErr != nil {
-			return "", nil, goErr
+		err = sendAndVerifyMsgSend(app, mapper, msg, ctx, []crypto.PrivKey{fromAcc.PrivKey}, handler)
+		if err != nil {
+			return "", nil, err
 		}
 		event("bank/sendAndVerifyMsgSend/ok")
 
 		return action, nil, nil
 	}
 }
 
+func createSingleInputSendMsg(r *rand.Rand, ctx sdk.Context, accs []simulation.Account, mapper auth.AccountMapper) (fromAcc simulation.Account, action string, msg bank.MsgSend, abort bool) {
+	fromAcc = simulation.RandomAcc(r, accs)
+	toAcc := simulation.RandomAcc(r, accs)
+	// Disallow sending money to yourself
+	for {
+		if !fromAcc.PubKey.Equals(toAcc.PubKey) {
+			break
+		}
+		toAcc = simulation.RandomAcc(r, accs)
+	}
+	toAddr := toAcc.Address
+	initFromCoins := mapper.GetAccount(ctx, fromAcc.Address).GetCoins()
+
+	if len(initFromCoins) == 0 {
+		return fromAcc, "skipping, no coins at all", msg, true
+	}
+
+	denomIndex := r.Intn(len(initFromCoins))
+	amt, goErr := randPositiveInt(r, initFromCoins[denomIndex].Amount)
+	if goErr != nil {
+		return fromAcc, "skipping bank send due to account having no coins of denomination " + initFromCoins[denomIndex].Denom, msg, true
+	}
+
+	action = fmt.Sprintf("%s is sending %s %s to %s",
+		fromAcc.Address.String(),
+		amt.String(),
+		initFromCoins[denomIndex].Denom,
+		toAddr.String(),
+	)
+
+	coins := sdk.Coins{{initFromCoins[denomIndex].Denom, amt}}
+	msg = bank.MsgSend{
+		Inputs:  []bank.Input{bank.NewInput(fromAcc.Address, coins)},
+		Outputs: []bank.Output{bank.NewOutput(toAddr, coins)},
+	}
+	return
+}
+
 // Sends and verifies the transition of a msg send. This fails if there are repeated inputs or outputs
-func sendAndVerifyMsgSend(app *baseapp.BaseApp, mapper auth.AccountMapper, msg bank.MsgSend, ctx sdk.Context, privkeys []crypto.PrivKey) error {
+// pass in handler as nil to handle txs, otherwise handle msgs
+func sendAndVerifyMsgSend(app *baseapp.BaseApp, mapper auth.AccountMapper, msg bank.MsgSend, ctx sdk.Context, privkeys []crypto.PrivKey, handler sdk.Handler) error {
```

Comment:
This pattern is a bit strange (handler nil/non-nil), for simulation it's not a big deal though.